components/compositing/largest_contentful_paint.rs

```diff
@@ -0,0 +1,91 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use base::cross_process_instant::CrossProcessInstant;
+use compositing_traits::largest_contentful_paint_candidate::{
+    LCPCandidate, LargestContentfulPaint,
+};
+use fnv::{FnvBuildHasher, FnvHashMap};
+use webrender_api::{Epoch, PipelineId};
+
+/// Holds the [`LargestContentfulPaintCalculator`] for each pipeline.
+#[derive(Default)]
+pub(crate) struct LargestContentfulPaintDetector {
+    lcp_calculators: FnvHashMap<PipelineId, LargestContentfulPaintCalculator>,
+}
+
+impl LargestContentfulPaintDetector {
+    pub const fn new() -> Self {
+        Self {
+            lcp_calculators: FnvHashMap::with_hasher(FnvBuildHasher::new()),
+        }
+    }
+
+    fn ensure_lcp_calculator(
+        &mut self,
+        pipeline_id: PipelineId,
+    ) -> &mut LargestContentfulPaintCalculator {
+        self.lcp_calculators
+            .entry(pipeline_id)
+            .or_insert(LargestContentfulPaintCalculator {
+                lcp_candidates: Vec::new(),
+                latest_lcp: None,
+            })
+    }
+
+    pub fn append_lcp_candidates(&mut self, pipeline_id: PipelineId, candidate: LCPCandidate) {
+        self.ensure_lcp_calculator(pipeline_id)
+            .lcp_candidates
+            .push(candidate);
+    }
+
+    pub fn calculate_largest_contentful_paint(
+        &mut self,
+        paint_time: CrossProcessInstant,
+        cur_epoch: Epoch,
+        pipeline_id: PipelineId,
+    ) -> Option<LargestContentfulPaint> {
+        self.ensure_lcp_calculator(pipeline_id)
+            .calculate_largest_contentful_paint(paint_time, cur_epoch)
```

Comment:
```suggestion
        match self.lcp_calculators.entry(pipeline_id) {
          OccupiedEntry(entry) => entry.get().calculate_largest_contentful_paint(),
          VacantEntry(_) => None,
        }
```
We shouldn't create a new calculator here.

Reply:
For the same pipeline_id, `fn ensure_lcp_calculator` dosen't create a new calculator.